solutions/usaco-1160.mdx

```diff
@@ -0,0 +1,153 @@
+---
+id: usaco-1160
+source: USACO Silver 2021 December
+title: Convoluted Intervals
+author: Aadit Ambadkar
+---
+
+[Official Analysis](http://usaco.org/current/data/sol_prob3_silver_dec21.html)
+
+## Alternate Solution
+An alternate solution to this problem is to use complementary counting. Consider a given value of $k$. 
+Count the number of pairs of intervals $i$ and $j$ which satisfy $a_i + a_j > k$ or $b_i + b_j < k$. It is clear that 
+any pair that satisfies $a_i + a_j > k$ cannot satisfy $b_i + b_j < k$, and vice versa. So, these two are 
+disjoint, and we can simply count the number of pairs which satisfy $a_i + a_j > k$ and add that to the 
+number of pairs which satisfy $b_i + b_j < k$. We subtract $n$ from this sum, and that gives the answer for
+that specific value of $k$. This solution is presented below.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+// Created by Aadit Ambadkar
+	
+#include<bits/stdc++.h>
+using namespace std;
+typedef long long ll;
+
+int main() {
+	int n, m; cin >> n >> m;
+	vector<pair<int, int>> v;
+	for (int i = 0; i < n; i++) {
+		int a, b; cin >> a >> b;
+		v.push_back({a, b});
+	}
+	for (int k = 0; k <= 2 * m; k++) {
+		ll acount = 0;
+		ll bcount = 0;
+		for (auto p : v) {
+			for (auto q : v) {
+				if (p.first + q.first > k) acount++;
+				else if (p.second + q.second < k) bcount++;
+			}
+		}
+		ll total = (ll) n * (ll) n;
+		cout << total - acount - bcount << "\n";
+	}
+}
+```
+</CPPSection>
+</LanguageSection>
+
+As the aboves stands, the solution runs in $\mathcal{O}(M\cdot N^2)$. This is enough to pass the first two
+test cases. To pass the first five, we make some optimizations. Create two arrays, `asum` and `bsum`.
+These arrays will store the number of pairs of intervals which sum to a given value, that is:
+
+`asum[x] = # of pairs i and j which satisfy a_i + a_j = x`
```

Comment:
`asum` is an unintuitive name
at least name it `l_sum` or `s_sum`

Reply:
How are those more intuitive than `asum`?

Reply:
what does `a` even mean
`l` or `s` would at least have some indication that it's `left` or `start`

Reply:
It's defined in the problem statement.

Reply:
i'm sure some people would think in terms of `a` and `b`, but `s` or `l` would just be easier imo

Reply:
... it is literally called a_i and b_i...